apps/elasticsearch.go

```diff
@@ -5,8 +5,101 @@ import (
 
 	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
 	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/otel"
 )
 
+type MetricsReceiverElasticsearch struct {
+	confgenerator.ConfigComponent          `yaml:",inline"`
+	confgenerator.MetricsReceiverShared    `yaml:",inline"`
+	confgenerator.MetricsReceiverSharedTLS `yaml:",inline"`
+
+	Endpoint string `yaml:"endpoint" validate:"omitempty,url,startswith=http:|startswith=https:"`
+
+	Username string `yaml:"username" validate:"required_with=Password"`
+	Password string `yaml:"password" validate:"required_with=Username"`
+
+	CollectJVMMetrics     *bool `yaml:"collect_jvm_metrics" validate:"omitempty"`
+	CollectClusterMetrics *bool `yaml:"collect_cluster_metrics" validate:"omitempty"`
+}
+
+const (
+	defaultElasticsearchEndpoint = "http://localhost:9200"
+)
+
+var (
+	defaultElasticsearchCollectClusterMetrics = true
+)
+
+func (r MetricsReceiverElasticsearch) Type() string {
+	return "elasticsearch"
+}
+
+func (r MetricsReceiverElasticsearch) Pipelines() []otel.Pipeline {
+	if r.Endpoint == "" {
+		r.Endpoint = defaultElasticsearchEndpoint
+	}
+
+	if r.CollectClusterMetrics == nil {
+		collectClusterMetrics := &defaultElasticsearchCollectClusterMetrics
+		r.CollectClusterMetrics = collectClusterMetrics
+	}
+
+	cfg := map[string]interface{}{
+		"collection_interval":  r.CollectionIntervalString(),
+		"endpoint":             r.Endpoint,
+		"username":             r.Username,
+		"password":             r.Password,
+		"nodes":                []string{"_local"},
+		"tls":                  r.TLSConfig(true),
+		"skip_cluster_metrics": !*r.CollectClusterMetrics,
+	}
+
+	if r.CollectJVMMetrics != nil && !*r.CollectJVMMetrics {
```

Comment:
Similar here, the logic around `CollectJVMMetrics` can also be moved to a shared util and be shared by Tomcat, Apache and ElasticSearch (or any other applications that need this flag), so that the behavior is consistent, and in the future we don't need to worry about details of this flag.

Reply:
Refactored into a shared struct. Because Elasticsearch isn't a JMX receiver, it doesn't share that extra logic it has to skip those metrics (it still uses this shared struct for consistency of flag naming and default values).